dietpi/dietpi-autostart

```diff
@@ -170,7 +170,9 @@ NB: You might need to grant the user permissions to run the chosen autostart opt
 
 				[[ ! -d ~$user/.config/cava ]] && mkdir -p ~$user/.config && cp /root/.config/cava ~$user/.config/cava && chown -R $user: ~$user/.config
 				[[ ! -f ~$user/cava.psf ]] && cp /root/cava.psf ~$user/cava.psf && chown $user: ~$user/cava.psf
-
+			elif (($ID_AUTOSTART == 2)); then
+				# File manager desktop icon
+				[[ -f '/usr/share/applications/pcmanfm.desktop' ]] && G_EXEC ln -sf /usr/share/applications/pcmanfm.desktop /root/Desktop/pcmanfm.desktop
```

Comment:
What was the idea here? Adding PCManFM desktop entry should still only be done within LXDE install step since other desktops have other file managers. Also one would want to have it regardless of autologin or LightDM login, manual `startx` call or VNC/XRDP, right?

What we could do, is similarly to how it's done in case of CAVA, looping through known desktop-related desktop entries `/root/Desktop/*.desktop` (only the ones which we link within any shared or individual desktop install step) and link those to the chosen users desktop dir, if there is no existing entry yet. So when one alters or removes entries from the root desktop, these changes are ported to new users as well, and if something is changed on a new users desktop, it is not overwritten, only missing entries are re-created.

Reply:
What if the user is actually using the root user, as a means to configure the system? Then that system would break down. I didn't do anything, partly because I was busy, but I also do not like that idea. How about just letting default applications be at `/usr/share/applications/`, and then doing, what you suggested?
Also the `pcfmanfm.desktop` icon will only be copied, if the file-manager is actually installed. 
What is done here, should also be done for this. 
https://github.com/MichaIng/DietPi/blob/d44e9aad4b3b1f10ef1adc0be14054ac4b1b9c14/dietpi/dietpi-software#L9157

Reply:
> What if the user is actually using the root user, as a means to configure the system? Then that system would break down.

The root user itself could be skipped for that loop, otherwise the linking would silently fail when linking a symlink to itself. But I agree it's not optimal, since user might want to setup root and regular users desktops differently, might clean up the root desktop or add additional admin entries, while this is not wanted for other users.

> Also the pcfmanfm.desktop icon will only be copied, if the file-manager is actually installed.

Ah that is true, only if the application file exists, so that is fine then. It would however missing when starting the desktop only manually, via VNC or automated boot into desktop. But actually, I just recognised that the file manager is one of the three default panel icons. Actually with this, it doesn't need to be placed onto the desktop as well, does it? Also every /usr/share/applications entry is automatically in the panel menu, so from there desktop icons can be easily created, if required. 

Making every /usr/share/applications a desktop icon automatically would be too much, I think, since it's already 16 alone for DietPi scripts and GL/ES info/testing. So a few selected ones, and newly installed desktop applications/games generally make sense. But:

> a login-manager could login any user, and I wouldn't be able to check, which user logged in, and if they already have the desktop icons in place.

That is true, no change to know it in this case...

> Furthermore, how do I detect changes to /usr/share/applications (which shouldn't happen?) ?

My idea was to loop through the same list selected entries where currently dietpi-software places a link to root's desktop and apply this to the chosen autologin user. But indeed when changing the user back and forth, manually removed icons would be recreated. I mean that could even be okay, usually on an SBC you're not regularly change an autologin desktop user, and if so, having 5-10 desktop icons re-created is probably acceptable. But the issue with LightDM remains.

Not sure if I understood your idea with the systemd unit. A service can disable itself e.g. as ExecStartPost step. But it would then run only for one user.
____________
A different very simple approach:
```
/etc/skel/Desktop/htop.desktop
```
The skeleton directory is copied automatically in any new users home directory. This includes system users as long as via adduser it is not explicitly unset or via useradd explicitly set. But the desktop entries are symlinks and the only other left file `.gtk-bookmarks` is tiny, so no large overhead, if a system user gets those additional files into his home.
On software installs, we can then additionally loop through user homes and add the one desktop icon that is to be installed. dietpi-software/config/launcher btw only makes sense on root's home since those require root permissions anyway. That way it is guaranteed that every file is only created once and every now and existing user get's once the entries which we think are helpful enough. What do you think?

Reply:
Ok, so going upon that, we could create a global user service in `/etc/systemd/system` , which starts on login, if there is no `~/.gtk-bookmarks`, and then the service copies all the files (as the action of the unit).
This would solve the problem, of copying files, where they shouldn't be, since systemd afaik only starts on users, which can log in (so actual human users, and not service users).

Reply:
I'm still not sure how to implement, since there is no systemd target for an actual user login 🤔. The systemd boot sequence ends with the TTY1 (what you see on the locally attached monitor, if there is one) login prompt, i.e. the `getty@tty1.service`. What happens after that, cannot be automated via systemd services, from all I know. But there are shell and desktop startup script/applications:
- `/etc/profile`, `/etc/profile.d/*.sh` and `~/.profile` are loaded whenever a user logs into a login shell, e.g. via local console or SSH.
- `/etc/bash.bashrc`, `/etc/bashrc.d/*.sh|.bash` and `~/.bashrc` are loaded whenever a user opens a bash shell.
- But what fits best here: `/etc/xdg/autostart/*.desktop` or `~/.config/autostart/*.desktop`: https://specifications.freedesktop.org/autostart-spec/autostart-spec-latest.html
    A desktop autostart entry indeed starts up when the desktop environment is loaded. And a systemd wide entry can be disabled by creating a same named file in the user autostart dir which contains `Hidden=true`.
    So in theory we could make that desktop entry creation script a system-wide autostart entry which automatically disables itself on user-level, once files have been copied. Indeed that would rule out files being created for system users or even regular users which are just not used to log into desktop environments. That way we can indeed skip pre-creating any desktop entries or configs for any user, including root. And we don't need to check for a flag file or such. We just need to loop through our "known" desktop entries/configs, and link/copy them, if they exist on system-level and do not exist already on user-level (should not happen, but theoretically a user could manually pre-create it's desktop icons, e.g. from a home dir backup or such). For `.gtk-bookmarks`, we'd need to download and store it somewhere on the system, e.g. `/var/lib/dietpi/dietpi-software/desktops/.gtk-bookmarks` at install.

Reply:
I think `user.slice` is created, when a user logs in.
Also, if a service is a user service, it won't get run(or even be tried to run), until a user actually loggs in.
You have the option of deciding, weather we do a `.desktop` autostart, or a systemd service.

Reply:
Good find and good to know! But it is active as well on console logins, so that desktop autostart entry would allow to easily apply things on actual desktop sessions only. I'll implement this.

Reply:
You can set services to start after `graphical.target` (which would be desktop) , or  `multiuser.target` (which would be console).
Also should we close this pr, or do you just want to take over from here (In this option, you should assign yourself to this pr)?

Reply:
`graphical.target` is also reached when no one logs into a desktop, even when the system is completely headless:
```
2021-02-02 12:12:08 root@micha:/tmp# systemctl status graphical.target
● graphical.target - Graphical Interface
     Loaded: loaded (/lib/systemd/system/graphical.target; static)
     Active: active since Mon 2021-02-01 21:55:01 CET; 14h ago
       Docs: man:systemd.special(7)

Feb 01 21:55:01 my.domain.org systemd[1]: Reached target Graphical Interface.
```
What gives this target a purpose is the `display-manager.service`, if it is present and enabled. This can be theoretically a desktop autostart, when manually setup that way, but is usually the LightDM (or other diplay/login manager) login prompt, but not the login itself. So using that target, the way I currently think of, would fire our script at every boot, either when LightDM prompt appears or, if not enabled, when the console login appears. In the latter case `graphical.target` and `multiuser.target` are reached directly one after another, without any meaning:
```
Jan 30 00:31:45 my.domain.org systemd[1]: Reached target Multi-User System.
Jan 30 00:31:45 my.domain.org systemd[1]: Reached target Graphical Interface.
```
But maybe I'm too bound in the way I think of systemd, and to be true I have not much experiences with user-level services and slices. So if you want to play with it and come up with a PR (or even this one), I would be happy to migrate things to a systemd-based solution, which would allow likely more fine grained features and ties into the general boot and login order, which is all based on systemd.

If it's fine for you, I'd take it over within this PR, which contains already a lot of desktop config migrations to system-level, which is awesome. It's only the parts that need to be installed into each users home that cause a little headache here to do right 😉.

Reply:
Ok, then go ahead. As for `graphical.target` and `multi-user.target` being reached seperately, I am also scratching my head. I think it may have something to do with having no `display-manager.service`. Another reason to abandon/depractice #4007 .
I also have not worked with user-level systemd services, so let's just cheat our way through.

Reply:
Yes, btw on that topic, related to Steam and Box86 testing, I also dealt a bit with this topic and found that there is actually no reliable method anymore to run an X server as non-root user. There is the `xserver-xorg-legacy` package which attempts to auto-detect the type of graphics driver and, if required, automatically executes X as root, which seems to have been the old default behaviour. However, it sometimes fails to detect it correctly (tested a bit up and down on VMs) and will then still fail to run X. No system group membership fixes this as it's somewhere deeper in the /dev/* directory structure. Granting render/video/tty group memberships works however for most modern graphics drivers. Last resort is to configure `xserver-xorg-legacy` a way that it always executes X as root, so does not do the test, but that breaks all security efforts behind the new attempt.

But it works fantastic to run LightDM and configure it to autologin as a certain user. It can be even configured to run any other X client application, so Steam or Amiberry or Kodi etc can all be launched through LightDM via any user without the need to deal with any sorts of graphics permissions.

I'm not able to port the way how LightDM achieves this, since it's quite complicated, but it's not necessary since LightDM is anyway mostly installed together with desktops.

I'll not implement this with v6.35 anymore, all the Box86 and Steam up and down testing costed too much time, but next release for sure.